lab3van/util-linux/schedutils/ionice.1
.TH IONICE 1 "July 2011" "util-linux" "User Commands"
.SH NAME
ionice \- set or get process I/O scheduling class and priority
.SH SYNOPSIS
.B ionice
.RB [ \-c
.IR class ]
.RB [ \-n
.IR level ]
.RB [ \-t ]
.B \-p
.IR PID ...
.br
.B ionice
.RB [ \-c
.IR class ]
.RB [ \-n
.IR level ]
.RB [ \-t ]
<<<<<<< HEAD
=======
.B \-P
.IR PGID ...
.br
.B ionice
.RB [ \-c
.IR class ]
.RB [ \-n
.IR level ]
.RB [ \-t ]
.B \-u
.IR UID ...
.br
.B ionice
.RB [ \-c
.IR class ]
.RB [ \-n
.IR level ]
.RB [ \-t ]
>>>>>>> master-vanilla
.IR "command " [ argument ...]
.SH DESCRIPTION
This program sets or gets the I/O scheduling class and priority for a program.
If no arguments or just \fB\-p\fR is given, \fBionice\fR will query the current
I/O scheduling class and priority for that process.

When \fIcommand\fR is given,
.B ionice
will run this command with the given arguments.
If no \fIclass\fR is specified, then
.I command
will be executed with the "best-effort" scheduling class.  The default
priority level is 4.

As of this writing, a process can be in one of three scheduling classes:
.IP "\fBIdle\fP"
A program running with idle I/O priority will only get disk time when no other
program has asked for disk I/O for a defined grace period.  The impact of an
idle I/O process on normal system activity should be zero.  This scheduling
class does not take a priority argument.  Presently, this scheduling class
is permitted for an ordinary user (since kernel 2.6.25).
.IP "\fBBest-effort\fP"
This is the effective scheduling class for any process that has not asked for
a specific I/O priority.
This class takes a priority argument from \fI0-7\fR, with a lower
number being higher priority.  Programs running at the same best-effort
priority are served in a round-robin fashion.

Note that before kernel 2.6.26 a process that has not asked for an I/O priority
formally uses "\fBnone\fP" as scheduling class, but the I/O scheduler will treat
such processes as if it were in the best-effort class.  The priority within the
best-effort class will be dynamically derived from the CPU nice level of the
process: io_priority = (cpu_nice + 20) / 5.

For kernels after 2.6.26 with the CFQ I/O scheduler, a process that has not asked
for an I/O priority inherits its CPU scheduling class.  The I/O priority is derived
from the CPU nice level of the process (same as before kernel 2.6.26).

.IP "\fBRealtime\fP"
The RT scheduling class is given first access to the disk, regardless of
what else is going on in the system.  Thus the RT class needs to be used with
some care, as it can starve other processes.  As with the best-effort class,
8 priority levels are defined denoting how big a time slice a given process
will receive on each scheduling window.  This scheduling class is not
permitted for an ordinary (i.e., non-root) user.
.SH OPTIONS
.TP
.BR \-c , " \-\-class " \fIclass\fR
Specify the name or number of the scheduling class to use; \fI0\fR for none,
\fI1\fR for realtime, \fI2\fR for best-effort, \fI3\fR for idle.
.TP
.BR \-n , " \-\-classdata " \fIlevel\fR
Specify the scheduling class data.  This only has an effect if the class
accepts an argument.  For realtime and best-effort, \fI0-7\fR are valid data
(priority levels).
.TP
.BR \-p , " \-\-pid " \fIPID\fR...
Specify the process IDs of running processes for which to get or set the
scheduling parameters.
.TP
<<<<<<< HEAD
=======
.BR \-P , " \-\-pgid " \fIPGID\fR...
Specify the process group IDs of running processes for which to get or set the
scheduling parameters.
.TP
>>>>>>> master-vanilla
.BR \-t , " \-\-ignore"
Ignore failure to set the requested priority.  If \fIcommand\fR was specified,
run it even in case it was not possible to set the desired scheduling priority,
which can happen due to insufficient privileges or an old kernel version.
.TP
.BR \-h , " \-\-help"
<<<<<<< HEAD
Display help and exit.
=======
Display help text and exit.
.TP
.BR \-u , " \-\-uid " \fIUID\fR...
Specify the user IDs of running processes for which to get or set the
scheduling parameters.
>>>>>>> master-vanilla
.TP
.BR \-V , " \-\-version"
Display version information and exit.
.SH EXAMPLES
.LP
.TP 7
# \fBionice\fP -c 3 -p 89
.TP 7
Sets process with PID 89 as an idle I/O process.
.TP 7
# \fBionice\fP -c 2 -n 0 bash
.TP 7
Runs 'bash' as a best-effort program with highest priority.
.TP 7
# \fBionice\fP -p 89 91
.TP 7
Prints the class and priority of the processes with PID 89 and 91.
.SH NOTES
Linux supports I/O scheduling priorities and classes since 2.6.13 with the CFQ
I/O scheduler.
.SH AUTHORS
.nf
Jens Axboe <jens@axboe.dk>
Karel Zak <kzak@redhat.com>
.fi
.SH AVAILABILITY
The ionice command is part of the util-linux package and is available from
ftp://ftp.kernel.org/pub/linux/utils/util-linux/.


debian/changelog
util-linux (2.27.160-1) unstable; urgency=low

  * Initial release (Closes: #nnnn)  <nnnn is the bug number of your ITP>

 -- roma <newangel88@mail.ru>  Thu, 24 Dec 2015 13:07:17 +0400


lab3van/util-linux/schedutils/ionice.c
/*
 * ionice: set or get process io scheduling class and priority
 *
 * Copyright (C) 2005 Jens Axboe <jens@axboe.dk>
 *
 * Released under the terms of the GNU General Public License version 2
 *
 */
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <getopt.h>
#include <unistd.h>
#include <sys/syscall.h>
#include <ctype.h>

#include "nls.h"
#include "strutils.h"
#include "c.h"
<<<<<<< HEAD
=======
#include "closestream.h"
>>>>>>> master-vanilla

static int tolerant;

static inline int ioprio_set(int which, int who, int ioprio)
{
	return syscall(SYS_ioprio_set, which, who, ioprio);
}

static inline int ioprio_get(int which, int who)
{
	return syscall(SYS_ioprio_get, which, who);
}

enum {
	IOPRIO_CLASS_NONE,
	IOPRIO_CLASS_RT,
	IOPRIO_CLASS_BE,
	IOPRIO_CLASS_IDLE,
};

enum {
	IOPRIO_WHO_PROCESS = 1,
	IOPRIO_WHO_PGRP,
	IOPRIO_WHO_USER,
};

#define IOPRIO_CLASS_SHIFT	(13)
#define IOPRIO_PRIO_MASK	((1UL << IOPRIO_CLASS_SHIFT) - 1)

#define IOPRIO_PRIO_CLASS(mask)	((mask) >> IOPRIO_CLASS_SHIFT)
#define IOPRIO_PRIO_DATA(mask)	((mask) & IOPRIO_PRIO_MASK)
#define IOPRIO_PRIO_VALUE(class, data)	(((class) << IOPRIO_CLASS_SHIFT) | data)

const char *to_prio[] = {
	[IOPRIO_CLASS_NONE] = "none",
	[IOPRIO_CLASS_RT]   = "realtime",
	[IOPRIO_CLASS_BE]   = "best-effort",
	[IOPRIO_CLASS_IDLE] = "idle"
};

static int parse_ioclass(const char *str)
{
	size_t i;

	for (i = 0; i < ARRAY_SIZE(to_prio); i++)
		if (!strcasecmp(str, to_prio[i]))
			return i;
	return -1;
}

<<<<<<< HEAD
static void ioprio_print(int pid)
{
	int ioprio = ioprio_get(IOPRIO_WHO_PROCESS, pid);
=======
static void ioprio_print(int pid, int who)
{
	int ioprio = ioprio_get(who, pid);
>>>>>>> master-vanilla

	if (ioprio == -1)
		err(EXIT_FAILURE, _("ioprio_get failed"));
	else {
		int ioclass = IOPRIO_PRIO_CLASS(ioprio);
		const char *name = _("unknown");

<<<<<<< HEAD
		if (ioclass > 0 && (size_t) ioclass < ARRAY_SIZE(to_prio))
			name = to_prio[ioclass];

		if (ioclass != IOPRIO_CLASS_IDLE)
			printf("%s: prio %lu\n", name,
					IOPRIO_PRIO_DATA(ioprio));
=======
		if (ioclass >= 0 && (size_t) ioclass < ARRAY_SIZE(to_prio))
			name = to_prio[ioclass];

		if (ioclass != IOPRIO_CLASS_IDLE)
			printf(_("%s: prio %lu\n"), name,
			       IOPRIO_PRIO_DATA(ioprio));
>>>>>>> master-vanilla
		else
			printf("%s\n", name);
	}
}

<<<<<<< HEAD
static void ioprio_setpid(pid_t pid, int ioclass, int data)
{
	int rc = ioprio_set(IOPRIO_WHO_PROCESS, pid,
=======
static void ioprio_setid(int which, int ioclass, int data, int who)
{
	int rc = ioprio_set(who, which,
>>>>>>> master-vanilla
			    IOPRIO_PRIO_VALUE(ioclass, data));

	if (rc == -1 && !tolerant)
		err(EXIT_FAILURE, _("ioprio_set failed"));
}

static void __attribute__ ((__noreturn__)) usage(FILE * out)
{
<<<<<<< HEAD
	fprintf(out,
	       _("\n"
		 "%1$s - sets or gets process io scheduling class and priority.\n"
		 "\n"
		 "Usage:\n"
		 "  %1$s [OPTION] -p PID [PID...]\n"
		 "  %1$s [OPTION] COMMAND\n"
		 "\n"
		 "Options:\n"
		 "  -c, --class <class>   scheduling class name or number\n"
		 "                           0: none, 1: realtime, 2: best-effort, 3: idle\n"
		 "  -n, --classdata <num> scheduling class data\n"
		 "                           0-7 for realtime and best-effort classes\n"
		 "  -p, --pid=PID         view or modify already running process\n"
		 "  -t, --ignore          ignore failures\n"
		 "  -V, --version         output version information and exit\n"
		 "  -h, --help            display this help and exit\n\n"),
		program_invocation_short_name);
=======
	fputs(USAGE_HEADER, out);
	fprintf(out,  _(" %1$s [options] -p <pid>...\n"
			" %1$s [options] -P <pgid>...\n"
			" %1$s [options] -u <uid>...\n"
			" %1$s [options] <command>\n"), program_invocation_short_name);

	fputs(USAGE_SEPARATOR, out);
	fputs(_("Show or change the I/O-scheduling class and priority of a process.\n"), out);

	fputs(USAGE_OPTIONS, out);
	fputs(_(" -c, --class <class>    name or number of scheduling class,\n"
		"                          0: none, 1: realtime, 2: best-effort, 3: idle\n"), out);
	fputs(_(" -n, --classdata <num>  priority (0..7) in the specified scheduling class,\n"
		"                          only for the realtime and best-effort classes\n"), out);
	fputs(_(" -p, --pid <pid>...     act on these already running processes\n"), out);
	fputs(_(" -P, --pgid <pgrp>...   act on already running processes in these groups\n"), out);
	fputs(_(" -t, --ignore           ignore failures\n"), out);
	fputs(_(" -u, --uid <uid>...     act on already running processes owned by these users\n"), out);

	fputs(USAGE_SEPARATOR, out);
	fputs(USAGE_HELP, out);
	fputs(USAGE_VERSION, out);

	fprintf(out, USAGE_MAN_TAIL("ionice(1)"));
>>>>>>> master-vanilla

	exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

int main(int argc, char **argv)
{
	int data = 4, set = 0, ioclass = IOPRIO_CLASS_BE, c;
<<<<<<< HEAD
	pid_t pid = 0;
=======
	int which = 0, who = 0;
	const char *invalid_msg = NULL;
>>>>>>> master-vanilla

	static const struct option longopts[] = {
		{ "classdata", required_argument, NULL, 'n' },
		{ "class",     required_argument, NULL, 'c' },
		{ "help",      no_argument,       NULL, 'h' },
		{ "ignore",    no_argument,       NULL, 't' },
		{ "pid",       required_argument, NULL, 'p' },
<<<<<<< HEAD
=======
		{ "pgid",      required_argument, NULL, 'P' },
		{ "uid",       required_argument, NULL, 'u' },
>>>>>>> master-vanilla
		{ "version",   no_argument,       NULL, 'V' },
		{ NULL, 0, NULL, 0 }
	};

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
<<<<<<< HEAD

	while ((c = getopt_long(argc, argv, "+n:c:p:tVh", longopts, NULL)) != EOF)
		switch (c) {
		case 'n':
			data = strtol_or_err(optarg, _("failed to parse class data"));
=======
	atexit(close_stdout);

	while ((c = getopt_long(argc, argv, "+n:c:p:P:u:tVh", longopts, NULL)) != EOF)
		switch (c) {
		case 'n':
			data = strtos32_or_err(optarg, _("invalid class data argument"));
>>>>>>> master-vanilla
			set |= 1;
			break;
		case 'c':
			if (isdigit(*optarg))
<<<<<<< HEAD
				ioclass = strtol_or_err(optarg,
						_("failed to parse class"));
=======
				ioclass = strtos32_or_err(optarg,
						_("invalid class argument"));
>>>>>>> master-vanilla
			else {
				ioclass = parse_ioclass(optarg);
				if (ioclass < 0)
					errx(EXIT_FAILURE,
						_("unknown scheduling class: '%s'"),
						optarg);
			}
			set |= 2;
			break;
		case 'p':
<<<<<<< HEAD
			pid = strtol_or_err(optarg, _("failed to parse pid"));
=======
			if (who)
				errx(EXIT_FAILURE,
				     _("can handle only one of pid, pgid or uid at once"));
			invalid_msg = _("invalid PID argument");
			which = strtos32_or_err(optarg, invalid_msg);
			who = IOPRIO_WHO_PROCESS;
			break;
		case 'P':
			if (who)
				errx(EXIT_FAILURE,
				     _("can handle only one of pid, pgid or uid at once"));
			invalid_msg = _("invalid PGID argument");
			which = strtos32_or_err(optarg, invalid_msg);
			who = IOPRIO_WHO_PGRP;
			break;
		case 'u':
			if (who)
				errx(EXIT_FAILURE,
				     _("can handle only one of pid, pgid or uid at once"));
			invalid_msg = _("invalid UID argument");
			which = strtos32_or_err(optarg, invalid_msg);
			who = IOPRIO_WHO_USER;
>>>>>>> master-vanilla
			break;
		case 't':
			tolerant = 1;
			break;
		case 'V':
<<<<<<< HEAD
			printf(_("%s from %s\n"),
				program_invocation_short_name, PACKAGE_STRING);
=======
			printf(UTIL_LINUX_VERSION);
>>>>>>> master-vanilla
			return EXIT_SUCCESS;
		case 'h':
			usage(stdout);
		default:
			usage(stderr);
		}

	switch (ioclass) {
		case IOPRIO_CLASS_NONE:
			if ((set & 1) && !tolerant)
				warnx(_("ignoring given class data for none class"));
			data = 0;
			break;
		case IOPRIO_CLASS_RT:
		case IOPRIO_CLASS_BE:
			break;
		case IOPRIO_CLASS_IDLE:
			if ((set & 1) && !tolerant)
				warnx(_("ignoring given class data for idle class"));
			data = 7;
			break;
		default:
			if (!tolerant)
				warnx(_("unknown prio class %d"), ioclass);
			break;
	}

<<<<<<< HEAD
	if (!set && !pid && optind == argc)
		/*
		 * ionice without options, print the current ioprio
		 */
		ioprio_print(0);

	else if (!set && pid) {
		/*
		 * ionice -p PID [PID ...]
		 */
		ioprio_print(pid);

		for(; argv[optind]; ++optind) {
			pid = strtol_or_err(argv[optind], _("failed to parse pid"));
			ioprio_print(pid);
		}
	} else if (set && pid) {
		/*
		 * ionice -c CLASS -p PID [PID ...]
		 */
		ioprio_setpid(pid, ioclass, data);

		for(; argv[optind]; ++optind) {
			pid = strtol_or_err(argv[optind], _("failed to parse pid"));
			ioprio_setpid(pid, ioclass, data);
=======
	if (!set && !which && optind == argc)
		/*
		 * ionice without options, print the current ioprio
		 */
		ioprio_print(0, IOPRIO_WHO_PROCESS);
	else if (!set && who) {
		/*
		 * ionice -p|-P|-u ID [ID ...]
		 */
		ioprio_print(which, who);

		for(; argv[optind]; ++optind) {
			which = strtos32_or_err(argv[optind], invalid_msg);
			ioprio_print(which, who);
		}
	} else if (set && who) {
		/*
		 * ionice -c CLASS -p|-P|-u ID [ID ...]
		 */
		ioprio_setid(which, ioclass, data, who);

		for(; argv[optind]; ++optind) {
			which = strtos32_or_err(argv[optind], invalid_msg);
			ioprio_setid(which, ioclass, data, who);
>>>>>>> master-vanilla
		}
	} else if (argv[optind]) {
		/*
		 * ionice [-c CLASS] COMMAND
		 */
<<<<<<< HEAD
		ioprio_setpid(0, ioclass, data);
		execvp(argv[optind], &argv[optind]);
		err(EXIT_FAILURE, _("executing %s failed"), argv[optind]);
=======
		ioprio_setid(0, ioclass, data, IOPRIO_WHO_PROCESS);
		execvp(argv[optind], &argv[optind]);
		err(EXIT_FAILURE, _("failed to execute %s"), argv[optind]);
>>>>>>> master-vanilla
	} else
		usage(stderr);


	return EXIT_SUCCESS;
}
